### PR TITLE
[CoreML ML Program] support acclerators selector

### PR DESCRIPTION
**Pull request by wejoncy**
### Description
For no, CoreML only support run mlmodels on CPU/ALL, However, sometimes CPU_GPU would be faster a lot.

We support the option to select different hardware to boost performance in this PR.

### CODE REVIEW

File: include/onnxruntime/core/providers/coreml/coreml_provider_factory.h [hunk 1]
```diff
@@ -28,8 +24,14 @@ enum COREMLFlags {
...
 
+  // Using CPU only in CoreML EP, this may decrease the perf but will provide
+  // reference output value without precision loss, which is useful for validation
+  COREML_FLAG_USE_CPU_ONLY = 0x010,
+  COREML_FLAG_USE_CPUAndGPU = 0x020,
+  COREML_FLAG_USE_CPUAndNeuralEngine = 0x040,
+  COREML_FLAG_USE_UnitAll = 0x080,
   // Create an MLProgram. By default it will create a NeuralNetwork model. Requires Core ML 5 or later.
-  COREML_FLAG_CREATE_MLPROGRAM = 0x010,
+  COREML_FLAG_CREATE_MLPROGRAM = 0x100,
```
Comment (skottmckay) [inline include/onnxruntime/core/providers/coreml/coreml_provider_factory.h:34]: We shouldn't change any existing flag as that will break any current usage of the EP.

Could we maybe just add a new 'no ANE' flag if the problem we're trying to solve is that large data sizes run less efficiently if CoreML attempts to use the ANE?

Comment (wejoncy) [reply include/onnxruntime/core/providers/coreml/coreml_provider_factory.h:34]: Yes, we could.

But the flag's value shouldn't be really matter given that we are using this enum variable. 
Please Correct me if I am wrong.

Comment (wejoncy) [reply include/onnxruntime/core/providers/coreml/coreml_provider_factory.h:34]: Besides, 'no ANE' wouldn't always give the best performance on different Mac.

File: include/onnxruntime/core/providers/coreml/coreml_provider_factory.h [hunk 2]
```diff
@@ -31,9 +31,11 @@ enum COREMLFlags {
   // Create an MLProgram. By default it will create a NeuralNetwork model. Requires Core ML 5 or later.
   COREML_FLAG_CREATE_MLPROGRAM = 0x010,
 
+  // Exclude ANE as sometimes this decrease performance
+  COREML_FLAG_USE_CPUAndGPU = 0x020,
```
Comment (edgchen1) [inline include/onnxruntime/core/providers/coreml/coreml_provider_factory.h:35]: nit: a name like `COREML_FLAG_USE_CPU_AND_GPU_ONLY` would be more consistent with the other names.

Comment (edgchen1) [reply include/onnxruntime/core/providers/coreml/coreml_provider_factory.h:35]: with this new flag, what does it mean if a user specifies both `COREML_FLAG_USE_CPU_ONLY` and `COREML_FLAG_USE_CPUAndGPU`? should we treat that as an error?

File: include/onnxruntime/core/providers/coreml/coreml_provider_factory.h (hunk 1 above)
Comment (edgchen1) [reply include/onnxruntime/core/providers/coreml/coreml_provider_factory.h]: even if users use the enum, if they have an older version of the header, their old enum value might now mean something else. that would be unexpected.

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 3]
```diff
@@ -395,9 +395,14 @@ Status Predict(const std::unordered_map<std::string, OnnxTensorData>& inputs,
       compiled_model_path_ = [compileUrl path];
 
       MLModelConfiguration* config = [[MLModelConfiguration alloc] init];
-      config.computeUnits = (coreml_flags_ & COREML_FLAG_USE_CPU_ONLY)
-                                ? MLComputeUnitsCPUOnly
-                                : MLComputeUnitsAll;
+      if (coreml_flags_ & COREML_FLAG_USE_CPU_ONLY) {
+        config.computeUnits = MLComputeUnitsCPUOnly;
+      } else if (coreml_flags_ & COREML_FLAG_USE_CPUAndGPU) {
+        config.computeUnits = MLComputeUnitsCPUAndGPU;
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:401]: should we also consider handling [MLComputeUnitsCPUAndNeuralEngine](https://developer.apple.com/documentation/coreml/mlcomputeunits/cpuandneuralengine?language=objc)?

Comment (skottmckay) [reply onnxruntime/core/providers/coreml/model/model.mm:401]: The problem this is addressing is that CoreML choosing to use ANE can result in worse perf for a model with large data sizes. Really CoreML compute unit selection should do a better job, but we don't control that.

I'm not aware of a scenario where excluding GPU would be of benefit, so I suggested we don't handle that as there's no known use case.

File: include/onnxruntime/core/providers/coreml/coreml_provider_factory.h (hunk 2 above)
Comment (wejoncy) [reply include/onnxruntime/core/providers/coreml/coreml_provider_factory.h:35]: CoreMLtools have the name as CPUAndGPU CPUAndANE CPUONLY ALL
https://github.com/apple/coremltools/blob/275c7e2c3ac44a23cd013a8b8a5f90c1a9991a46/coremlpython/CoreMLPython.mm#L144

Should we align the name with them？

Comment (wejoncy) [reply include/onnxruntime/core/providers/coreml/coreml_provider_factory.h:35]: "CPU_AND_GPU ONLY " sounds a little weird. "CPU ONLY", "GPU ONLY" is more meanningful.

Comment (wejoncy) [reply include/onnxruntime/core/providers/coreml/coreml_provider_factory.h:35]: > with this new flag, what does it mean if a user specifies both `COREML_FLAG_USE_CPU_ONLY` and `COREML_FLAG_USE_CPUAndGPU`? should we treat that as an error?

Yeah, I think so. Will fix it.

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 4]
```diff
@@ -395,9 +395,24 @@ Status Predict(const std::unordered_map<std::string, OnnxTensorData>& inputs,
...
-                                ? MLComputeUnitsCPUOnly
-                                : MLComputeUnitsAll;
+      uint32_t device_flags = coreml_flags_ & (COREML_FLAG_USE_CPU_ONLY | COREML_FLAG_USE_CPU_AND_GPU);
+      // check if only one flag is set
+      if (device_flags & (device_flags - 1)) {
+        // multiple device options selected
+        return ORT_MAKE_STATUS(
+            ONNXRUNTIME, FAIL,
+            "Multiple device options selected, "
+            "you should use one of the following options: COREML_FLAG_USE_CPU_ONLY or COREML_FLAG_USE_CPU_AND_GPU");
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:405]:
```suggestion
            "you should use at most one of the following options: COREML_FLAG_USE_CPU_ONLY or COREML_FLAG_USE_CPU_AND_GPU");
```

File: java/src/main/java/ai/onnxruntime/providers/CoreMLFlags.java [hunk 5]
```diff
@@ -25,7 +25,9 @@ public enum CoreMLFlags implements OrtFlags {
    * Create an MLProgram. By default it will create a NeuralNetwork model. Requires Core ML 5 or
    * later.
    */
-  CREATE_MLPROGRAM(16); // COREML_FLAG_CREATE_MLPROGRAM(0x010)
+  CREATE_MLPROGRAM(16), // COREML_FLAG_CREATE_MLPROGRAM(0x010)
+  /** exclude ANE */
+  CPU_And_GPU(32); // COREML_FLAG_USE_CPU_AND_GPU(0x020)
```
Comment (edgchen1) [inline java/src/main/java/ai/onnxruntime/providers/CoreMLFlags.java:30]:
```suggestion
  CPU_AND_GPU(32); // COREML_FLAG_USE_CPU_AND_GPU(0x020)
```

File: include/onnxruntime/core/providers/coreml/coreml_provider_factory.h [hunk 6]
```diff
@@ -31,9 +31,14 @@ enum COREMLFlags {
   // Create an MLProgram. By default it will create a NeuralNetwork model. Requires Core ML 5 or later.
   COREML_FLAG_CREATE_MLPROGRAM = 0x010,
 
+  // Exclude ANE as sometimes this decrease performance
+  // https://developer.apple.com/documentation/coreml/mlcomputeunits?language=objc
+  // there are four compute units:
+  // MLComputeUnitsCPUAndNeuralEngine|MLComputeUnitsCPUAndGPU|MLComputeUnitsCPUOnly|MLComputeUnitsAll
+  COREML_FLAG_USE_CPU_AND_GPU = 0x020,
```
Comment (edgchen1) [inline include/onnxruntime/core/providers/coreml/coreml_provider_factory.h:38]: if we're calling it `USE_CPU_AND_GPU` and not calling it `EXCLUDE_ANE`, I'd suggest documenting it as such. i.e., something like "use CPU and GPU for computation in CoreML". you can still mention that using this to exclude the ANE may increase performance.

File: include/onnxruntime/core/providers/coreml/coreml_provider_factory.h (hunk 2 above)
Comment (edgchen1) [reply include/onnxruntime/core/providers/coreml/coreml_provider_factory.h]: aligning with the coreml names seems reasonable.

File: include/onnxruntime/core/providers/coreml/coreml_provider_factory.h (hunk 6 above)
Comment (wejoncy) [reply include/onnxruntime/core/providers/coreml/coreml_provider_factory.h:38]: For different models, running on different hardwares, the performance varies. It's hard to say which option is the best. To try with different CU and pick the best one which shows better performance is a good way.

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 7]
```diff
@@ -395,9 +395,24 @@ Status Predict(const std::unordered_map<std::string, OnnxTensorData>& inputs,
...
-                                : MLComputeUnitsAll;
+      uint32_t device_flags = coreml_flags_ & (COREML_FLAG_USE_CPU_ONLY | COREML_FLAG_USE_CPU_AND_GPU);
+      // check if only one flag is set
+      if (device_flags & (device_flags - 1)) {
+        // multiple device options selected
+        return ORT_MAKE_STATUS(
+            ONNXRUNTIME, FAIL,
+            "Multiple device options selected, "
+            "you should use at most one of the following options: COREML_FLAG_USE_CPU_ONLY or COREML_FLAG_USE_CPU_AND_GPU");
+      }
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/model/model.mm:406]: Can we do this check in the C++ code when creating the EP as that is the point where the user options are being passed in? 

It's also easier to maintain/test as that code can run on any platform.

Comment (wejoncy) [reply onnxruntime/core/providers/coreml/model/model.mm:406]: Sounds good

Comment (wejoncy) [reply onnxruntime/core/providers/coreml/model/model.mm]: Moved to the EP construct function.

File: include/onnxruntime/core/providers/coreml/coreml_provider_factory.h (hunk 6 above)
Comment (edgchen1) [reply include/onnxruntime/core/providers/coreml/coreml_provider_factory.h:38]: sure, you can document the performance implications however you see fit.

what I'm trying to suggest is to be consistent with the name, "COREML_FLAG_USE_CPU_AND_GPU" and the documentation for it, "Exclude ANE". I think either changing the name to "COREML_FLAG_EXCLUDE_ANE" or changing the documentation to "Use CPU and GPU" would make it more consistent.

File: onnxruntime/core/providers/coreml/coreml_execution_provider.cc [hunk 8]
```diff
@@ -32,6 +32,13 @@ CoreMLExecutionProvider::CoreMLExecutionProvider(uint32_t coreml_flags)
     LOGS_DEFAULT(ERROR) << "CoreML EP is not supported on this platform.";
   }
 
+  // check if only one flag is set
+  if ((coreml_flags & COREML_FLAG_USE_CPU_ONLY) && (coreml_flags & COREML_FLAG_USE_CPU_AND_GPU)) {
+    // multiple device options selected
+    LOGS_DEFAULT(ERROR) << "Multiple device options selected, you should use at most one of the following options:"
+                        << "COREML_FLAG_USE_CPU_ONLY or COREML_FLAG_USE_CPU_AND_GPU or not set";
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply onnxruntime/core/providers/coreml/coreml_execution_provider.cc]: Agree

**Review by skottmckay — APPROVED** [review]

File: include/onnxruntime/core/providers/coreml/coreml_provider_factory.h (hunk 6 above)
Comment (wejoncy) [reply include/onnxruntime/core/providers/coreml/coreml_provider_factory.h:38]: sounds good. Will fix it in the next PR.

**Pull request merged**